cloudimg/ms_azure.py

```diff
@@ -0,0 +1,513 @@
+        storage_account_name (str): The storage account name for auth.
+        storage_account_key (str): The storage account key for auth.
+    """
+    # The duration of generated SAS URIs
+    SAS_EXPIRY = int(os.getenv("CLOUDIMG_AZURE_SAS_EXPIRY", 365 * 3))
+    # The max concurrency for blob upload
+    UPLOAD_MAX_CONCURRENCY = int(os.getenv(
+                                "CLOUDIMG_AZURE_UPLOAD_MAX_CONCURRENCY",
+                                16,
+                             ))
+    # The interval in seconds between showing upload progress in logs
+    UPLOAD_LOG_INTERVAL = int(os.getenv("CLOUDIMG_AZURE_LOG_INTERVAL", 15))
+
+    def __init__(self, blob_service_client, storage_account_name,
+                 storage_account_key, **kwargs):
+
+        self.blob_service_client = blob_service_client
+        self._account_name = storage_account_name
+        self._account_key = storage_account_key
+        self._upload_progress = UploadProgress(0, 0)
+        self._upload_last_log = self._upload_progress.last_seen
+
+        super(AzureService, self).__init__(**kwargs)
+
+    @classmethod
+    def from_connection_string(cls, connection_str):
+        """
+        Factory for AzureService using the storage connection string for auth.
+
+        Args:
+            connection_str (str): The storage connection string for auth.
+        """
+        values = _get_values_from_connection_string(connection_str)
+        blob_service_client = BlobServiceClient.from_connection_string(
+                                  connection_str
+                              )
+        return cls(
+            blob_service_client=blob_service_client,
+            storage_account_name=values['AccountName'],
+            storage_account_key=values['AccountKey'],
+        )
+
+    @classmethod
+    def from_storage_account(cls, account_name, account_key):
+        """
+        Factory for AzureService using the storage account name and key
+        for auth.
+
+        Args:
+            account_name (str): The storage account name.
+            account_key (str): The storage account key.
+        Returns:
+            AzureService object
+        """
+        sas_token = generate_account_sas(
+            account_name=account_name,
+            account_key=account_key,
+            resource_types=ResourceTypes(
+                service=True,
+                container=True,
+                object=True
+            ),
+            permission=AccountSasPermissions(
+                read=True,
+                write=True,
+                list=True,
+                add=True,
+                create=True,
+                update=True,
+                process=True,
+                tag=True,
+                filter_by_tags=True,
+            ),
+            expiry=datetime.utcnow() + timedelta(days=cls.SAS_EXPIRY)
+        )
+        blob_service_client = BlobServiceClient(
+            account_url="https://%s.blob.core.windows.net" % account_name,
```

Comment:
Should the hardcoded instances of blob.core.windows.net perhaps be configurable, even if just by an env var (like the other magic constants)?

In AWS-related code we often make the endpoint URL configurable so that we are able to test the code realistically in a sandbox by running https://github.com/localstack/localstack.  I'm not familiar with Azure but it seems there are a couple of similar things there (Azurite, Azure Storage Emulator) and it probably wouldn't be possible to connect this code with those tools if it's hardcoded to always use these hostnames.

Reply:
I've changed it to allow passing the value to an env var.